File: ports/lua/lib/push.lua
-- push.lua: sprite-tool push — Push local file or directory to a sprite.

local sprite = require("sprite")

local M = {}

local function usage()
    print([[Usage: sprite-tool push <local-path> <remote-path> [sprite-name]

Examples:
  sprite-tool push ./file.txt /home/sprite/file.txt
  sprite-tool push ./mydir /home/sprite/mydir
  sprite-tool push ./file.txt /tmp/file.txt ember-red-hawk]])
    os.exit(1)
end

--- Check if a path exists (file or directory).
local function path_exists(path)
    local fh = io.open(path, "r")
    if fh then
        fh:close()
        return true
    end
    -- Also check if it's a directory
    local ok = os.execute("test -e " .. require("config").shell_quote(path))
    return ok == true or ok == 0
end

--- Check if a path is a directory.
local function is_dir(path)
    local ok = os.execute("test -d " .. require("config").shell_quote(path))
    return ok == true or ok == 0
end

--- Main entry point for the push subcommand.
function M.run(args)
    if #args < 2 then usage() end

    local local_path  = args[1]
    local remote_path = args[2]
    local sprite_name = args[3] or ""

    if not path_exists(local_path) then
        io.stderr:write("Error: " .. local_path .. " does not exist\n")
        os.exit(1)
    end

    if is_dir(local_path) then
        print("Pushing directory: " .. local_path .. " -> " .. remote_path)
        sprite.push_dir_strip(sprite_name, local_path, remote_path, false)
    else
        print("Pushing file: " .. local_path .. " -> " .. remote_path)
        sprite.push_file(sprite_name, local_path, remote_path, false)
    end

    print("Done.")
    return 0
end

return M

File: ports/haskell/src/Push.hs
-- | sprite-tool push: Push local file or directory to a sprite.
module Push (run) where

import System.Directory (doesDirectoryExist, doesFileExist)
import System.Exit      (exitFailure, ExitCode(..))
import System.FilePath  (takeDirectory)
import System.IO        (hPutStrLn, stderr, hGetContents, hPutStr, hFlush, hClose)
import System.Process   ( createProcess, proc, waitForProcess
                        , CreateProcess(..), StdStream(..) )

-- | Print usage and exit.
usage :: IO a
usage = do
    putStrLn $ unlines
        [ "Usage: sprite-tool push <local-path> <remote-path> [sprite-name]"
        , ""
        , "Examples:"
        , "  sprite-tool push ./file.txt /home/sprite/file.txt"
        , "  sprite-tool push ./mydir /home/sprite/mydir"
        , "  sprite-tool push ./file.txt /tmp/file.txt ember-red-hawk"
        ]
    exitFailure

-- | Execute the push subcommand.
run :: [String] -> IO ()
run args
    | length args < 2 = usage
    | otherwise = do
        let localPath  = args !! 0
            remotePath = args !! 1
            spriteName = if length args > 2 then args !! 2 else ""
            sArgs      = if null spriteName then [] else ["-s", spriteName]

        -- Check local path exists
        fileExists <- doesFileExist localPath
        dirExists  <- doesDirectoryExist localPath
        if not fileExists && not dirExists
            then do
                hPutStrLn stderr $ "Error: " ++ localPath ++ " does not exist"
                exitFailure
            else return ()

        if dirExists
            then pushDirectory localPath remotePath sArgs
            else pushSingleFile localPath remotePath sArgs

        putStrLn "Done."

-- | Push a directory via tar pipe.
pushDirectory :: String -> String -> [String] -> IO ()
pushDirectory localPath remotePath sArgs = do
    putStrLn $ "Pushing directory: " ++ localPath ++ " -> " ++ remotePath
    let parent = takeDirectory localPath
        base   = takeBaseName' localPath

    -- tar czf - -C parent base | sprite exec ... bash -c "mkdir -p DEST && tar xzf - -C DEST --strip-components=1"
    (_, Just tarOut, _, tarPh) <- createProcess
        (proc "tar" ["czf", "-", "-C", parent, base])
        { std_out = CreatePipe }

    let spriteCmd = ["sprite", "exec"] ++ sArgs ++
            ["bash", "-c", "mkdir -p '" ++ remotePath ++ "' && tar xzf - -C '" ++ remotePath ++ "' --strip-components=1"]
    (Just spriteIn, _, _, spritePh) <- createProcess
        (proc (head spriteCmd) (tail spriteCmd))
        { std_in = CreatePipe }

    -- Pipe tar output to sprite stdin
    tarContents <- hGetContents tarOut
    hPutStr spriteIn tarContents
    hFlush spriteIn
    hClose spriteIn

    spriteEc <- waitForProcess spritePh
    _ <- waitForProcess tarPh

    case spriteEc of
        ExitSuccess   -> return ()
        ExitFailure _ -> do
            hPutStrLn stderr "Error: push failed"
            exitFailure

-- | Push a single file via stdin pipe.
pushSingleFile :: String -> String -> [String] -> IO ()
pushSingleFile localPath remotePath sArgs = do
    putStrLn $ "Pushing file: " ++ localPath ++ " -> " ++ remotePath
    let remoteDir = takeDirectory remotePath

    let spriteCmd = ["sprite", "exec"] ++ sArgs ++
            ["bash", "-c", "mkdir -p '" ++ remoteDir ++ "' && cat > '" ++ remotePath ++ "'"]

    fileContents <- readFile localPath
    (Just hin, _, _, ph) <- createProcess
        (proc (head spriteCmd) (tail spriteCmd))
        { std_in = CreatePipe }

    hPutStr hin fileContents
    hFlush hin
    hClose hin

    ec <- waitForProcess ph
    case ec of
        ExitSuccess   -> return ()
        ExitFailure _ -> do
            hPutStrLn stderr "Error: push failed"
            exitFailure

-- | Extract directory basename (last path component, preserving extension).
takeBaseName' :: String -> String
takeBaseName' p =
    let stripped = if not (null p) && last p == '/' then init p else p
    in reverse $ takeWhile (/= '/') (reverse stripped)
